crates/dslab-iaas/readme.md

```diff
@@ -0,0 +1,87 @@
+# DSLab IaaS
+
+## Overview
+
+A library for cloud simulations creation.
+
+It can be used by cloud computing academic studies to implement user specific simulations. The simulator contains a set of standard components which allow developers to focus on specific system design issues to be investigated, without concerning the low-level details of simulation. At the same time, any component can be easily extended or replaced by the user to meet their behavior requirements. As the result, the library supports simple, easy-to-understand simulations and also grants the access to low-level details.
+
+The library enables modelling of different cloud infrastructure items such as physical machines (or [HostManager](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/core/host_manager.rs)), virtual machines ([VirtualMachine](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/core/vm.rs)) and other components of resource allocation and management. The simulation processes a chain of discrete events and their handlers brought from [dslab-core](https://github.com/osukhoroslov/dslab/tree/main/crates/dslab-core) crate. Any component implements the handler interface and use events to interact with others.
+
+The simulation is managed via [CloudSimulation](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/simulation.rs#L25), which provides an interface to spawn new physical machines, submit virtual machines and migrate them, add new user defined components etc. As long as simulation consists of events ordered by timestamp, it is possible to run simulation for a few steps or time period. 
+
+Basic examples of library usage can be found in [examples](https://github.com/osukhoroslov/dslab/blob/main/examples/iaas/src/main.rs) block.
+
+## Virtual machines allocation
+
+Virtual machines can be submitted to cloud cluster via scheduler component. The scheduler selects the most suitable physical host among all possible at the moment. The library supports multiple schedulers, thus a two-factor commit approach is used for virtual machine allocation. Conflicts are resolved by a centralized database called [PlacementStore](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/core/placement_store.rs#L20). After an allocation is confirmed, an allocation request is sent to the selected host.
```

Comment:
```suggestion
Virtual machines can be submitted to cloud resource pool via [Scheduler](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/core/scheduler.rs) component. The scheduler processes VM allocation events and selects the physical machine (aka host) for running VM. It stores a local copy of resource pool state, which includes information of current resource allocations on each host. Scheduler can also access information about current hosts’ load from the monitoring component. The actual VM placement decision is delegated to the user-defined
[VM placement algorithm](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/core/vm_placement_algorithm.rs).

The library supports simulating a cloud with multiple schedulers that concurrently process allocation requests. Since each scheduler operates using its own, possibly outdated resource pool state, the schedulers’ decisions may produce conflicts. These conflicts are resolved by a centralized database called [PlacementStore](https://github.com/osukhoroslov/dslab/blob/main/crates/dslab-iaas/src/core/placement_store.rs) (PS). Each scheduler sends
its placement decisions to the PS. If not conflicts are detected, the allocation is committed and an allocation request is sent to the selected host.
```